Add ARM64 node_e2e test

ike-ma opened the pull request:
Run node_e2e serial tests on GCP T2A (arm64) machine

* Test locally with the following command

```
KUBE_BUILD_FOR_ARM64=true ARTIFACTS="${TMPDIR}" JENKINS_GCE_SSH_PRIVATE_KEY_FILE="${SSH_KEY}" \
  kubetest   --up   --test   --provider=gce   \
  --deployment=node --gcp-project="${PROJECT}"   --gcp-zone="${ZONE}" \
  "--node-args=--image-config-file="${IMAGE_CONFIG_OUT}   \
  '--node-test-args="" --kubelet-flags=""'   --node-tests=true   \
  '--test_args=--nodes=1 --focus="\[Serial\]" \
  --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]"'   \
  '--timeout=300m' 2>&1 | tee -i "${TMPDIR}/build-log-$(date +%Y%m%d.%M.%S.%3N).txt"
```

* Test Result: PASS

```
Ran 34 of 390 Specs in 1619.273 seconds
SUCCESS! -- 34 Passed | 0 Failed | 0 Pending | 356 Skipped
PASS
```


Cross-ref: https://github.com/kubernetes/kubernetes/pull/117017  (Setup node_e2e to support ARM64)

ike-ma commented:
/cc @bobbypage

dims commented:
/ok-to-test

dims commented on config/jobs/kubernetes/sig-node/node-kubelet.yaml:
```diff
@@ -120,6 +120,45 @@ periodics:
+          - --scenario=kubernetes_e2e
+          - --
+          - --deployment=node
+          - --env=KUBE_BUILD_FOR_ARM64=true
```
@ike-ma where is this used? `KUBE_BUILD_FOR_ARM64`

ike-ma replied:
I was using it to control whether to build for ARM64 or AMD64: https://github.com/kubernetes/kubernetes/pull/117017/files/c7c8c7fe388079adf3a24e0d465dd35afea15357#diff-4e140fc3b0b901a6cdad3f45948687ebd709753a4c6b48a52e876a1fec856247

mmiranda96 replied:
Is this still relevant? It seems to be removed from kubernetes/kubernetes#117017

bobbypage replied:
+1, this should no longer be needed?

ike-ma replied:
This has been removed

ike-ma commented:
@bobbypage

ike-ma commented:
/assign @bobbypage

SergeyKanzhelev commented:
/assign @mmiranda96

mmiranda96 commented on jobs/e2e_node/arm/image-config-serial.yaml:
```diff
@@ -0,0 +1,10 @@
+# `gcloud compute --project <to-project> disks create <image name> --image=https://www.googleapis.com/compute/v1/projects/<from-project>/global/images/<image-name>`
+# `gcloud compute --project <to-project> images create <image-name> --source-disk=<image-name>`
+images:
+  ubuntu:
```
Do you plan to use COS as well? We can add both configs here.

ike-ma replied:
Currently only Ubuntu is in scope, will add COS after testing separately.

mmiranda96 commented on jobs/e2e_node/arm/image-config-serial.yaml:
```diff
@@ -0,0 +1,10 @@
+# `gcloud compute --project <to-project> images create <image-name> --source-disk=<image-name>`
+images:
+  ubuntu:
+    image: ubuntu-gke-2204-1-24-arm64-v20230217
```
I suggest that we use `image-family` instead, otherwise we'd need to update this job constantly.

ike-ma replied:
Done. Updated.

mmiranda96 commented on jobs/e2e_node/arm/init.yaml:
```diff
@@ -0,0 +1,28 @@
+runcmd:
+  - echo "Test run from /tmp folder, remounting it"
+  - mount /tmp /tmp -o remount,exec,suid
+
```
Not sure if having empty lines across a YAML list is supported.

ike-ma replied:
The empty lines will be ignored by the parser. I have tested it without any issue.

mmiranda96 commented:
/retest

mmiranda96 commented:
/assign @ndixita

tzneal commented:
/cc

bobbypage commented on config/jobs/kubernetes/sig-node/node-kubelet.yaml:
```diff
@@ -120,6 +120,45 @@ periodics:
+    description: "Run serial node e2e tests on ARM64 environment on Ubuntu"
+  labels:
+    preset-service-account: "true"
+    preset-k8s-ssh: "true"
```
This will likely need 

```
preset-dind-enabled: "true"
```

because ARM build needs to use dockerized build so we need to enable docker-in-docker.

ike-ma replied:
Thanks for the pointer, added.

bobbypage commented on config/jobs/kubernetes/sig-node/node-kubelet.yaml:
```diff
@@ -120,6 +120,45 @@ periodics:
+          - --deployment=node
+          - --env=KUBE_BUILD_FOR_ARM64=true
+          - --gcp-zone=us-central1-a
+          - --node-args=--image-config-file=/workspace/test-infra/jobs/e2e_node/arm/image-config.yaml
```
This will likely need the new flags introduced in https://github.dev/kubernetes/kubernetes/pull/117017, i.e.

```
--use-dockerized-build=true --target-build-arch=linux/arm64
```

ike-ma replied:
Done, updated with the latest command.

chendave replied:
seems like there is a typo in the config file, 

```
--image-config-file=/workspace/test-infra/jobs/e2e_node/arm/image-config.yaml
```

while it should 
```
--image-config-file=/workspace/test-infra/jobs/e2e_node/arm/image-config-serial.yaml
```

Can you confirm?

/hold

ike-ma replied:
Good catch. Named as `image-config.yaml` in my own test. Updated to match.

ike-ma commented:
Addressed all review comments, and updated the test config based on https://github.com/kubernetes/test-infra/issues/28888. 

Tested with 

```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --use-dockerized-build=true --target-build-arch=linux/arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml 

```

BenTheElder commented on config/jobs/kubernetes/sig-node/node-kubelet.yaml:
```diff
@@ -120,6 +120,49 @@ periodics:
+    containers:
+      - image: gcr.io/k8s-staging-test-infra/kubekins-e2e:v20230513-7e1db2f1bb-master
+        args:
+          - --repo=k8s.io/kubernetes=master
```
we should stop writing new jobs with bootstrap.py instead of `decorate: true`

these jobs have been logging a warning to migrate for _years_

this doesn't have to block adding the job, but please seriously consider migrating.

https://gist.github.com/dims/c1296f8ed42238baea0a5fcae45f4cf4

ike-ma replied:
Acknowledge, thanks for the gist pointer, will follow up in a separate PR.

BenTheElder reviewed (commented):
/approve
with a request to please switch to prow native cloning etc. (`decorate: true`) and stop using the bootstrap.py entrypoint script.

ike-ma commented:
Addressed review comments, and rebased to upstream/master

chendave commented:
/hold cancel
/lgtm
thanks!

chendave commented:
/test pull-test-infra-unit-test

SergeyKanzhelev commented:
Cool! Thank you for merging. It is failing: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-node-arm64-ubuntu-serial/1662289187593785344
